src/marker.rs

```diff
@@ -2,103 +2,161 @@
 use crate::impl_::not_send::NotSend;
 use crate::type_object::{PyTypeInfo, PyTypeObject};
 use crate::types::{PyAny, PyDict, PyModule, PyType};
+use crate::version::PythonVersionInfo;
 use crate::{ffi, AsPyPointer, FromPyPointer, IntoPyPointer, PyNativeType, PyObject, PyTryFrom};
+use std::convert::TryFrom;
 use std::ffi::{CStr, CString};
 use std::marker::PhantomData;
 use std::os::raw::c_int;
 
-/// Represents the major, minor, and patch (if any) versions of this interpreter.
+/// Types that are safe to access while the GIL is not held.
 ///
-/// See [Python::version].
-#[derive(Debug)]
-pub struct PythonVersionInfo<'py> {
-    /// Python major version (e.g. `3`).
-    pub major: u8,
-    /// Python minor version (e.g. `11`).
-    pub minor: u8,
-    /// Python patch version (e.g. `0`).
-    pub patch: u8,
-    /// Python version suffix, if applicable (e.g. `a0`).
-    pub suffix: Option<&'py str>,
-}
-
-impl<'py> PythonVersionInfo<'py> {
-    /// Parses a hard-coded Python interpreter version string (e.g. 3.9.0a4+).
-    ///
-    /// Panics if the string is ill-formatted.
-    fn from_str(version_number_str: &'py str) -> Self {
-        fn split_and_parse_number(version_part: &str) -> (u8, Option<&str>) {
-            match version_part.find(|c: char| !c.is_ascii_digit()) {
-                None => (version_part.parse().unwrap(), None),
-                Some(version_part_suffix_start) => {
-                    let (version_part, version_part_suffix) =
-                        version_part.split_at(version_part_suffix_start);
-                    (version_part.parse().unwrap(), Some(version_part_suffix))
-                }
-            }
-        }
-
-        let mut parts = version_number_str.split('.');
-        let major_str = parts.next().expect("Python major version missing");
-        let minor_str = parts.next().expect("Python minor version missing");
-        let patch_str = parts.next();
-        assert!(
-            parts.next().is_none(),
-            "Python version string has too many parts"
-        );
-
-        let major = major_str
-            .parse()
-            .expect("Python major version not an integer");
-        let (minor, suffix) = split_and_parse_number(minor_str);
-        if suffix.is_some() {
-            assert!(patch_str.is_none());
-            return PythonVersionInfo {
-                major,
-                minor,
-                patch: 0,
-                suffix,
-            };
-        }
-
-        let (patch, suffix) = patch_str.map(split_and_parse_number).unwrap_or_default();
-        PythonVersionInfo {
-            major,
-            minor,
-            patch,
-            suffix,
-        }
-    }
-}
+/// # Safety
+///
+/// The type must not carry borrowed Python references or, if it does, not allow access to them if
+/// the GIL is not held.
+///
+/// See the [module-level documentation](self) for more information.
+#[cfg(not(feature = "nightly"))]
+pub unsafe trait Ungil {}
```

Comment:
Bikesheds on the name are welcome.

Reply:
I just went for the shortest one I could think of - all the auto traits in std have really short names, so I figured it would fit. 😆

> cleanest in terms of description

IMO, the danger of names like these is that users might think they understand what it does but are wrong. So perhaps a more cryptic name is better.

Reply:
That's a fair point. I'm happy to keep the name `Ungil` 👍